File: python-legacy/src/elbencho_harness/report/compare.py
"""Compare report: overlay N run directories into one HTML.

A `bench run` produces a directory of result.json files (one per spec, possibly
many for a sweep). `bench compare A/ B/ C/` aligns those results across runs
and renders a multi-series report.

Alignment rules:
  - Each run is identified by its directory basename (or --label).
  - Specs match if they share workload_name AND target_name. Sweep axis values
    are surfaced as additional context in the table.
  - Baseline = first run argument (or --baseline). Deltas are pct change.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

from jinja2 import Environment, FileSystemLoader, select_autoescape

from ..results.schema import PhaseResult, Result
from ..results.store import list_results, read_manifest
from .charts import (
    axis_pretty_name,
    detect_common_axis,
    format_axis_value,
    overlay_with_toggle,
    render_figure_html,
    sweep_overlay,
)
from .render import TEMPLATES_DIR

import plotly.graph_objects as go


@dataclass
class ResultWithAxes:
    """One result paired with the sweep axis-values that produced it.

    Pairing happens by spec_dir, not spec_hash, because two sweep points can
    produce the same effective workload (and hence same spec_hash) when their
    overrides happen to equal the base workload's values.
    """

    result: Result
    axes: dict[str, Any] | None


@dataclass
class LoadedRun:
    """One run directory unrolled into its label + result list."""

    label: str
    run_dir: Path
    results: list[ResultWithAxes]


def load_run(run_dir: Path, *, label: str | None = None) -> LoadedRun:
    """Load a run directory in manifest order, pairing each result with its
    sweep axis-values."""
    rdir = run_dir.resolve()
    paired: list[ResultWithAxes] = []
    try:
        mf = read_manifest(rdir)
    except FileNotFoundError:
        mf = None
    if mf is not None:
        for entry in mf.run_specs:
            sd = rdir / entry["spec_dir"]
            rj = sd / "result.json"
            if not rj.is_file():
                continue
            try:
                result = Result.model_validate_json(rj.read_text())
            except Exception:
                continue
            paired.append(ResultWithAxes(result=result, axes=entry.get("axis_values")))
    else:
        # No manifest, fall back to listing result.json files in directory order.
        for r in list_results(rdir):
            paired.append(ResultWithAxes(result=r, axes=None))
    return LoadedRun(label=label or rdir.name, run_dir=rdir, results=paired)


def _spec_key(rwa: ResultWithAxes) -> tuple:
    """Stable identifier for matching specs across runs.

    Uses (target_name, workload_name, axis_values). Two specs with the same
    target+workload but different axis values won't collapse.
    """
    axis_tuple = tuple(sorted((rwa.axes or {}).items()))
    return (rwa.result.target.name, rwa.result.workload.name, axis_tuple)


def _primary_phase(r: Result) -> PhaseResult | None:
    return r.phases.get(r.primary_phase) or next(iter(r.phases.values()), None)


def _throughput_of(r: Result) -> float | None:
    p = _primary_phase(r)
    if p is None:
        return None
    v = p.throughput_mib_s_last or p.throughput_mib_s_first
    return float(v) if v is not None else None


def _iops_of(r: Result) -> float | None:
    p = _primary_phase(r)
    if p is None:
        return None
    v = p.iops_last or p.iops_first
    return float(v) if v is not None else None


def _avg_lat_of(r: Result) -> float | None:
    p = _primary_phase(r)
    if p is None or p.io_lat_us is None:
        return None
    return float(p.io_lat_us.avg) if p.io_lat_us.avg is not None else None


@dataclass
class CompareRow:
    spec_key: tuple
    target: str
    workload: str
    axis_label: str          # composite human label, e.g. "block_size=64KiB"
    axes: dict[str, Any] | None  # raw axis-values dict from the manifest
    per_run: dict[str, dict[str, float | None]]  # run_label -> {metric: value}

    def baseline_value(self, metric: str, baseline_label: str) -> float | None:
        return self.per_run.get(baseline_label, {}).get(metric)

    def delta_pct(self, metric: str, run_label: str, baseline_label: str) -> float | None:
        base = self.baseline_value(metric, baseline_label)
        v = self.per_run.get(run_label, {}).get(metric)
        if base in (None, 0) or v is None:
            return None
        return (v - base) / base * 100.0


def build_rows(runs: list[LoadedRun]) -> list[CompareRow]:
    """Align results across runs by spec_key. Returns one row per distinct spec."""
    keyed: dict[tuple, CompareRow] = {}
    for run in runs:
        for rwa in run.results:
            r = rwa.result
            key = _spec_key(rwa)
            row = keyed.get(key)
            if row is None:
                row = CompareRow(
                    spec_key=key,
                    target=r.target.name,
                    workload=r.workload.name,
                    axis_label=_format_axes(rwa.axes),
                    axes=rwa.axes,
                    per_run={},
                )
                keyed[key] = row
            row.per_run[run.label] = {
                "throughput_mib_s": _throughput_of(r),
                "iops": _iops_of(r),
                "avg_lat_us": _avg_lat_of(r),
                "duration_s": r.duration_s,
            }
    # Stable order: target, workload, then by the *numeric* axis value if one is
    # shared (so the chart x-axis sorts naturally for sweeps), else by label.
    common_axis = detect_common_axis([r.axes for r in keyed.values()])
    if common_axis is not None:
        return sorted(
            keyed.values(),
            key=lambda r: (r.target, r.workload, _numeric_axis_value(r.axes, common_axis)),
        )
    return sorted(
        keyed.values(), key=lambda r: (r.target, r.workload, r.axis_label)
    )


def _numeric_axis_value(axes: dict[str, Any] | None, axis: str) -> float:
    """Pull the sweep axis value as a number for sorting. Strings sort last."""
    if not axes:
        return float("inf")
    v = axes.get(axis)
    try:
        return float(v)
    except (TypeError, ValueError):
        return float("inf")


def _format_axes(axes: dict[str, Any] | None) -> str:
    if not axes:
        return ""
    parts = []
    for k, v in sorted(axes.items()):
        parts.append(f"{k}={v}")
    return ", ".join(parts)


# --- charts -----------------------------------------------------------------


def _chart_axis(rows: list[CompareRow]) -> tuple[list[str], str]:
    """Return (x_labels, x_axis_title) for a chart.

    When every row's axes dict has the same single axis (a clean 1-D sweep),
    use the axis value alone as the tick label ("64k", "4m", "16"). Otherwise
    fall back to the composite per-row label ("target·workload·axes").
    """
    common = detect_common_axis([r.axes for r in rows])
    if common is not None:
        x_labels = [format_axis_value(common, r.axes[common]) for r in rows]
        return x_labels, axis_pretty_name(common)
    # Multi-axis fallback: compose target + axes if any.
    fallback: list[str] = []
    for r in rows:
        s = f"{r.target}·{r.workload}"
        if r.axis_label:
            s += f"·{r.axis_label}"
        fallback.append(s)
    return fallback, ""


def _series_from_rows(
    runs: list[LoadedRun], rows: list[CompareRow], metric: str
) -> list[tuple[str, list[float | None]]]:
    return [
        (run.label, [r.per_run.get(run.label, {}).get(metric) for r in rows])
        for run in runs
    ]


def throughput_overlay(runs: list[LoadedRun]) -> go.Figure:
    """Throughput overlay with live Bar/Line toggle."""
    rows = build_rows(runs)
    x_labels, x_title = _chart_axis(rows)
    return sweep_overlay(
        title="Throughput",
        x_labels=x_labels,
        x_axis_title=x_title,
        y_title="MiB/s",
        series=_series_from_rows(runs, rows, "throughput_mib_s"),
        hover_unit="MiB/s",
    )


def iops_overlay(runs: list[LoadedRun]) -> go.Figure:
    """IOPS overlay with live Bar/Line toggle."""
    rows = build_rows(runs)
    x_labels, x_title = _chart_axis(rows)
    return sweep_overlay(
        title="IOPS",
        x_labels=x_labels,
        x_axis_title=x_title,
        y_title="IOPS",
        series=_series_from_rows(runs, rows, "iops"),
        hover_unit="IOPS",
    )


def latency_overlay(runs: list[LoadedRun]) -> go.Figure:
    """Average IO latency overlay with live Bar/Line toggle."""
    rows = build_rows(runs)
    x_labels, x_title = _chart_axis(rows)
    return sweep_overlay(
        title="Average IO latency",
        x_labels=x_labels,
        x_axis_title=x_title,
        y_title="µs",
        series=_series_from_rows(runs, rows, "avg_lat_us"),
        hover_unit="µs",
    )


# --- rendering --------------------------------------------------------------


def render_compare(
    runs: list[LoadedRun],
    out_path: Path,
    *,
    baseline_label: str | None = None,
    title: str = "elbencho-harness compare",
) -> Path:
    """Write a self-contained HTML compare report. Returns the path written."""
    if not runs:
        raise ValueError("render_compare requires at least one LoadedRun")
    baseline = baseline_label or runs[0].label
    rows = build_rows(runs)
    env = Environment(
        loader=FileSystemLoader(str(TEMPLATES_DIR)),
        autoescape=select_autoescape(["html", "htm", "xml"]),
        trim_blocks=True,
        lstrip_blocks=True,
    )
    template = env.get_template("compare.html.j2")
    n_series = len(runs)
    html = template.render(
        title=title,
        runs=runs,
        rows=rows,
        baseline=baseline,
        throughput_html=overlay_with_toggle(throughput_overlay(runs), n_series=n_series),
        iops_html=overlay_with_toggle(iops_overlay(runs), n_series=n_series),
        latency_html=overlay_with_toggle(latency_overlay(runs), n_series=n_series),
    )
    out_path.write_text(html, encoding="utf-8")
    return out_path

File: src/report/compare.rs
//! Compare report rendering (multi-run overlay).
//!
//! Loads N run directories from disk, aligns specs by (target, workload,
//! axis_values), emits one HTML with three overlay charts + a diff table
//! vs baseline.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use anyhow::{anyhow, Context, Result};
use askama::Template;
use indexmap::IndexMap;
use serde::Deserialize;

use super::charts::{axis_pretty_name, format_axis_value, overlay};
use crate::results::schema::Result as RunResult;

/// One result paired with the sweep axis values that produced it.
#[derive(Debug, Clone)]
pub struct ResultWithAxes {
    pub result: RunResult,
    pub axes: Option<IndexMap<String, serde_json::Value>>,
}

/// One loaded run directory ready for compare.
#[derive(Debug, Clone)]
pub struct LoadedRun {
    pub label: String,
    pub run_dir: PathBuf,
    pub results: Vec<ResultWithAxes>,
}

#[derive(Deserialize)]
struct ManifestFile {
    #[serde(default)]
    run_specs: Vec<ManifestEntry>,
}

#[derive(Deserialize)]
struct ManifestEntry {
    spec_dir: String,
    #[serde(default)]
    axis_values: Option<IndexMap<String, serde_json::Value>>,
}

/// Load a run directory: read manifest.json if present, then each spec's
/// result.json. Each result is paired with its sweep axis values (from the
/// manifest entry) so two distinct sweep points with the same effective
/// spec_hash stay distinct in the compare table.
pub fn load_run(run_dir: &Path, label: Option<&str>) -> Result<LoadedRun> {
    let rdir = run_dir
        .canonicalize()
        .with_context(|| format!("canonicalizing {}", run_dir.display()))?;
    let mut paired: Vec<ResultWithAxes> = Vec::new();

    let manifest_path = rdir.join("manifest.json");
    let manifest: Option<ManifestFile> = if manifest_path.is_file() {
        let text = std::fs::read_to_string(&manifest_path)?;
        serde_json::from_str(&text).ok()
    } else {
        None
    };

    if let Some(mf) = manifest {
        for entry in mf.run_specs {
            let rj = rdir.join(&entry.spec_dir).join("result.json");
            if let Ok(text) = std::fs::read_to_string(&rj) {
                if let Ok(result) = serde_json::from_str::<RunResult>(&text) {
                    paired.push(ResultWithAxes {
                        result,
                        axes: entry.axis_values,
                    });
                }
            }
        }
    } else {
        // No manifest, walk subdirs.
        for entry in std::fs::read_dir(&rdir)? {
            let entry = entry?;
            if !entry.file_type()?.is_dir() {
                continue;
            }
            let rj = entry.path().join("result.json");
            if let Ok(text) = std::fs::read_to_string(&rj) {
                if let Ok(result) = serde_json::from_str::<RunResult>(&text) {
                    paired.push(ResultWithAxes {
                        result,
                        axes: None,
                    });
                }
            }
        }
        // Stable order by dir name.
        paired.sort_by(|a, b| a.result.run_id.cmp(&b.result.run_id));
    }

    let label_str = label
        .map(String::from)
        .unwrap_or_else(|| rdir.file_name().unwrap_or_default().to_string_lossy().into_owned());
    Ok(LoadedRun {
        label: label_str,
        run_dir: rdir,
        results: paired,
    })
}

#[derive(Debug)]
struct CompareRow {
    #[allow(dead_code)]
    spec_key: (String, String, Vec<(String, serde_json::Value)>),
    target: String,
    workload: String,
    axis_label: String,
    axes: Option<IndexMap<String, serde_json::Value>>,
    per_run: HashMap<String, MetricSet>,
}

#[derive(Debug, Default, Clone, Copy)]
struct MetricSet {
    throughput: Option<f64>,
    iops: Option<f64>,
    avg_lat: Option<f64>,
}

fn primary_metrics(r: &RunResult) -> MetricSet {
    let phase = r
        .phases
        .get(&r.primary_phase)
        .or_else(|| r.phases.values().next());
    match phase {
        None => MetricSet::default(),
        Some(p) => MetricSet {
            throughput: p.throughput_mib_s_last.or(p.throughput_mib_s_first),
            iops: p.iops_last.or(p.iops_first),
            avg_lat: p.io_lat_us.avg,
        },
    }
}

fn spec_key(rwa: &ResultWithAxes) -> (String, String, Vec<(String, serde_json::Value)>) {
    let mut axes: Vec<(String, serde_json::Value)> = rwa
        .axes
        .as_ref()
        .map(|m| m.iter().map(|(k, v)| (k.clone(), v.clone())).collect())
        .unwrap_or_default();
    axes.sort_by(|a, b| a.0.cmp(&b.0));
    (rwa.result.target.name.clone(), rwa.result.workload.name.clone(), axes)
}

fn format_axes_label(axes: &Option<IndexMap<String, serde_json::Value>>) -> String {
    let Some(m) = axes else { return String::new() };
    let mut parts: Vec<(String, &serde_json::Value)> =
        m.iter().map(|(k, v)| (k.clone(), v)).collect();
    parts.sort_by(|a, b| a.0.cmp(&b.0));
    parts
        .into_iter()
        .map(|(k, v)| format!("{}={}", k, v))
        .collect::<Vec<_>>()
        .join(", ")
}

fn build_rows(runs: &[LoadedRun]) -> Vec<CompareRow> {
    let mut keyed: HashMap<(String, String, Vec<(String, serde_json::Value)>), CompareRow> =
        HashMap::new();
    for run in runs {
        for rwa in &run.results {
            let key = spec_key(rwa);
            let row = keyed.entry(key.clone()).or_insert_with(|| CompareRow {
                spec_key: key.clone(),
                target: rwa.result.target.name.clone(),
                workload: rwa.result.workload.name.clone(),
                axis_label: format_axes_label(&rwa.axes),
                axes: rwa.axes.clone(),
                per_run: HashMap::new(),
            });
            row.per_run.insert(run.label.clone(), primary_metrics(&rwa.result));
        }
    }
    let common_axis = detect_common_axis(&keyed.values().map(|r| r.axes.clone()).collect::<Vec<_>>());
    let mut rows: Vec<CompareRow> = keyed.into_values().collect();
    if let Some(axis) = common_axis {
        rows.sort_by(|a, b| {
            let av = numeric_axis_value(&a.axes, &axis);
            let bv = numeric_axis_value(&b.axes, &axis);
            (a.target.clone(), a.workload.clone(), av).partial_cmp(&(
                b.target.clone(),
                b.workload.clone(),
                bv,
            )).unwrap()
        });
    } else {
        rows.sort_by(|a, b| {
            (a.target.clone(), a.workload.clone(), a.axis_label.clone()).cmp(&(
                b.target.clone(),
                b.workload.clone(),
                b.axis_label.clone(),
            ))
        });
    }
    rows
}

fn detect_common_axis(axes_dicts: &[Option<IndexMap<String, serde_json::Value>>]) -> Option<String> {
    let mut axis_name: Option<String> = None;
    for d in axes_dicts {
        let Some(d) = d else { return None };
        if d.len() != 1 {
            return None;
        }
        let only = d.keys().next().unwrap().clone();
        match &axis_name {
            None => axis_name = Some(only),
            Some(existing) => {
                if existing != &only {
                    return None;
                }
            }
        }
    }
    axis_name
}

fn numeric_axis_value(
    axes: &Option<IndexMap<String, serde_json::Value>>,
    axis: &str,
) -> f64 {
    let Some(map) = axes else { return f64::INFINITY };
    let Some(v) = map.get(axis) else { return f64::INFINITY };
    match v {
        serde_json::Value::Number(n) => n.as_f64().unwrap_or(f64::INFINITY),
        serde_json::Value::String(s) => s.parse().unwrap_or(f64::INFINITY),
        _ => f64::INFINITY,
    }
}

fn axis_for_chart(rows: &[CompareRow]) -> (Vec<String>, &'static str) {
    let common = detect_common_axis(&rows.iter().map(|r| r.axes.clone()).collect::<Vec<_>>());
    if let Some(axis) = &common {
        let labels: Vec<String> = rows
            .iter()
            .map(|r| {
                let v = numeric_axis_value(&r.axes, axis);
                if v.is_finite() {
                    format_axis_value(axis, v as i64)
                } else {
                    "".into()
                }
            })
            .collect();
        (labels, axis_pretty_name(axis))
    } else {
        let labels: Vec<String> = rows
            .iter()
            .map(|r| {
                let mut s = format!("{}·{}", r.target, r.workload);
                if !r.axis_label.is_empty() {
                    s.push('·');
                    s.push_str(&r.axis_label);
                }
                s
            })
            .collect();
        (labels, "")
    }
}

#[derive(Template)]
#[template(path = "compare.html")]
struct ComparePage {
    title: String,
    baseline: String,
    run_count: usize,
    runs: Vec<RunPill>,
    rows: Vec<RowView>,
    run_columns: Vec<RunColumn>,
    throughput_json: String,
    iops_json: String,
    latency_json: String,
}

struct RunPill {
    label: String,
    spec_count: usize,
    run_dir: String,
}

struct RowView {
    target: String,
    workload: String,
    axis_label: String,
    cells: Vec<RowCell>,
}

struct RowCell {
    value_fmt: String,
    show_delta: bool,
    delta_fmt: String,
    delta_class: String,
}

struct RunColumn {
    label: String,
    show_delta: bool,
}

/// Write a multi-run compare HTML to out_path. Aligns specs by
/// (target, workload, axis values); diff column shows percentage change
/// vs the baseline run.
pub fn render_compare(
    runs: &[LoadedRun],
    out_path: &Path,
    baseline_label: Option<&str>,
    title: &str,
) -> Result<()> {
    if runs.is_empty() {
        return Err(anyhow!("render_compare requires at least one LoadedRun"));
    }
    let baseline = baseline_label
        .map(String::from)
        .unwrap_or_else(|| runs[0].label.clone());

    let rows = build_rows(runs);
    let (x_labels, x_title) = axis_for_chart(&rows);

    let series_for = |sel: fn(&MetricSet) -> Option<f64>| -> Vec<(String, Vec<Option<f64>>)> {
        runs.iter()
            .map(|run| {
                let ys: Vec<Option<f64>> = rows
                    .iter()
                    .map(|r| r.per_run.get(&run.label).and_then(|m| sel(m)))
                    .collect();
                (run.label.clone(), ys)
            })
            .collect()
    };

    let throughput_fig = overlay(
        "Throughput",
        x_title,
        "MiB/s",
        "MiB/s",
        &x_labels,
        &series_for(|m| m.throughput),
    );
    let iops_fig = overlay(
        "IOPS",
        x_title,
        "IOPS",
        "IOPS",
        &x_labels,
        &series_for(|m| m.iops),
    );
    let latency_fig = overlay(
        "Average IO latency",
        x_title,
        "µs",
        "µs",
        &x_labels,
        &series_for(|m| m.avg_lat),
    );

    let run_pills: Vec<RunPill> = runs
        .iter()
        .map(|r| RunPill {
            label: r.label.clone(),
            spec_count: r.results.len(),
            run_dir: r.run_dir.display().to_string(),
        })
        .collect();
    let run_columns: Vec<RunColumn> = runs
        .iter()
        .map(|r| RunColumn {
            label: r.label.clone(),
            show_delta: r.label != baseline,
        })
        .collect();

    // Build the diff-table rows.
    let row_views: Vec<RowView> = rows
        .iter()
        .map(|row| {
            let base_tput = row
                .per_run
                .get(&baseline)
                .and_then(|m| m.throughput);
            let cells: Vec<RowCell> = runs
                .iter()
                .map(|run| {
                    let m = row.per_run.get(&run.label);
                    let v = m.and_then(|m| m.throughput);
                    let value_fmt = match v {
                        None => "—".into(),
                        Some(v) => format!("{:.1}", v),
                    };
                    let show_delta = run.label != baseline;
                    let (delta_fmt, delta_class) = if show_delta {
                        match (base_tput, v) {
                            (Some(b), Some(v)) if b != 0.0 => {
                                let d = (v - b) / b * 100.0;
                                let class = if d > 0.5 {
                                    "delta-pos"
                                } else if d < -0.5 {
                                    "delta-neg"
                                } else {
                                    "delta-zero"
                                };
                                let label = if d > 0.0 {
                                    format!("+{:.1}%", d)
                                } else {
                                    format!("{:.1}%", d)
                                };
                                (label, class.into())
                            }
                            _ => ("—".into(), "dim".into()),
                        }
                    } else {
                        (String::new(), String::new())
                    };
                    RowCell {
                        value_fmt,
                        show_delta,
                        delta_fmt,
                        delta_class,
                    }
                })
                .collect();
            RowView {
                target: row.target.clone(),
                workload: row.workload.clone(),
                axis_label: row.axis_label.clone(),
                cells,
            }
        })
        .collect();

    let page = ComparePage {
        title: title.into(),
        baseline,
        run_count: runs.len(),
        runs: run_pills,
        rows: row_views,
        run_columns,
        throughput_json: serde_json::to_string(&throughput_fig)?,
        iops_json: serde_json::to_string(&iops_fig)?,
        latency_json: serde_json::to_string(&latency_fig)?,
    };
    let html = page.render()?;
    if let Some(parent) = out_path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(out_path, html)?;
    Ok(())
}
